exberry_adapter/bot/exberry_adapter_bot.py

```diff
@@ -99,6 +100,21 @@ def handle_order_request(event):
         logging.info(f'Received Create Order Request - {event}')
         order = event.cdata['details']
 
+        opt_fee_contract_id = order['optExchangeFee']
+        fee_amount = 0.0
+        if opt_fee_contract_id != None:
+            fee_deposit = client.find_by_id(opt_fee_contract_id)
+            fee_amount = fee_deposit.cdata['asset']['quantity']
+
+        fee_schedules = client.find_active(MARKETPLACE.FeeSchedule, {'provider': exchange_party})
+        for (_,schedule) in fee_schedules.items():
+            if schedule['currentFee']['amount'] > fee_amount:
+                return [exercise(event.cid,
+                            'RejectRequest', {
+                                'errorCode': '790',
+                                'errorMessage': 'Fee does not match exchange fee'}
+                        )]
```

Comment:
I'm not quite sure how I feel about having this business logic (or check) here in the Exberry adapter.

Could we not have this logic sit in Daml when a user creates an order ? Is the issue that the `customer` cannot see the `FeeSchedule` contract as it's not an `observer` on that contract nor has the signature from the `provider ?

Reply:
I would have preferred this to not be in the adapter as well, but you are correct that the issue is the `customer` cannot see the contract, so therefore would not be able to look it up in `CreateOrder` to perform this check. Do you have any other ideas on how to do this? I couldn't think of a way of doing it without multi-party submits or a trigger adding all Trading customers to the `FeeSchedule` `observers` list.

Reply:
I'd say a fee schedule contract exists for every customer. You could imagine different fee tiers depending on volume traded, so then you need to be able to associate a user with a specific fee schedule. Also, it would allow to set custom fees for some customers. The fee schedule would be something that is agreed to and created upon request or grant of the service. That would then also solve this problem as it would be visible to the user.

Reply:
@georg-da I agree that's the way to go - I discussed this with Eric initially and he requested we start with a single Fee Schedule per exchange and open it up to per customer at a later time. I could also see fees being different per listing.